internal/evaluator/builder.go

```diff
@@ -29,15 +29,11 @@ import (
 // Builder is a build-through cache of message evaluators keyed off the provided
 // descriptor.
 type Builder struct {
-	// TODO: (TCN-1708) based on benchmarks, about 50% of CPU time is spent obtaining a read
-	//  lock on this mutex. Ideally, this can be reworked to be thread-safe while
-	//  minimizing the need to obtain a lock.
-	mtx         sync.RWMutex
-	cache       map[protoreflect.MessageDescriptor]*message
+	cache       atomic.Pointer[map[protoreflect.MessageDescriptor]*message]
```

Comment:
since we're passing around the map into a bunch of functions, let's create a type for it (within internal):

```
type EvaluatorCache map[protoreflect.MessageDescriptor]*message
```

With the added benefit of being able to hang functionality off of it (like Clone) to make things a bit more ergonomic

Reply:
Added the type!

Reply:
Renamed to `MessageCache` to avoid stutter lint.